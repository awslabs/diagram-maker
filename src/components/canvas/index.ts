export { default as Canvas, CanvasProps } from './Canvas';
