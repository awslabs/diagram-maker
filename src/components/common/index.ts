export { default as ComposeView } from './ComposeView';
