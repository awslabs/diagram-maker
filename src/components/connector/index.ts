export {
  default as Connector, ConnectorProps, ConnectorType
} from './Connector';
