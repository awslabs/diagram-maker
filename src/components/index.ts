export { Canvas, CanvasProps } from './Canvas';
export { default as ComposeView } from './common/ComposeView';
export {
  default as Connector,
  ConnectorProps,
  ConnectorType
} from './connector/Connector';
export {
  default as ContextMenu,
  ContextMenuProps
} from './contextMenu/ContextMenu';
export { Node, NodeProps, PotentialNode, PotentialNodeProps } from './node';
export { Panel, PanelProps } from './panel';
export { SelectionMarquee, SelectionMarqueeProps } from './selectionMarquee';
export { View, ViewProps } from './view/View';
export { getConnectedView, ConnectedViewOwnProps } from './view/ConnectedView';
export { Workspace, WorkspaceProps } from './Workspace';

export * from './edge/EdgeCurve';
export { default as Edge, EdgeProps } from './edge/Edge';
export { default as EdgeBadge, EdgeBadgeProps } from './edge/EdgeBadge';
export {
  default as PotentialEdge,
  PotentialEdgeProps
} from './edge/PotentialEdge';

export {
  default as GridPattern,
  GridPatternProps
} from './canvas/patterns/GridPattern';

export {
  default as DotPattern,
  DotPatternProps
} from './canvas/patterns/DotPattern';

export { render, destroy } from './renderUtils';
