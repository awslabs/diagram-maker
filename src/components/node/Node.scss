.dm-node, .dm-potential-node {
  position: absolute;
  .dm-content {
    position: relative;
    height: 100%;
    z-index: 1;
    box-sizing: border-box;
  }
}

.dm-potential-node {
  z-index: 8000;
}
