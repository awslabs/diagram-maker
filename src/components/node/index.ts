export { default as Node, NodeProps } from './Node';
export { default as PotentialNode, PotentialNodeProps } from './PotentialNode';
