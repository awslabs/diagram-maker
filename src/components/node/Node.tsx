import * as Preact from 'preact';

import { ComposeView } from 'diagramMaker/components/common';
import {
  Connector, ConnectorProps, ConnectorType
} from 'diagramMaker/components/connector';
import {
  BoundRenderCallback, ConnectorPlacement, ConnectorPlacementType, DestroyCallback,
  TypeForVisibleConnectorTypes, VisibleConnectorTypes
} from 'diagramMaker/service/ConfigService';
import { DiagramMakerComponentsType } from 'diagramMaker/service/ui/types';
import { DiagramMakerNode } from 'diagramMaker/state/types';

import './Node.scss';

export interface NodeProps<NodeType> {
  connectorPlacement?: ConnectorPlacementType;
  renderCallback: BoundRenderCallback;
  destroyCallback: DestroyCallback;
  diagramMakerNode: DiagramMakerNode<NodeType>;
  visibleConnectorTypes?: TypeForVisibleConnectorTypes;
}

export default class Node<NodeType> extends Preact.Component<NodeProps<NodeType>, {}> {

  public render(): JSX.Element {
    const { diagramMakerData, id } = this.props.diagramMakerNode;
    const { x, y } = diagramMakerData.position;
    const { width, height } = diagramMakerData.size;
    const transform = `translate3d(${x}px, ${y}px, 0)`;
    const { renderCallback, destroyCallback } = this.props;

    return (
      <div
        className="dm-node"
        style={{ width, height, transform }}
        data-id={id}
        data-type={DiagramMakerComponentsType.NODE}
        data-event-target={true}
        data-draggable={true}
      >
        <ComposeView
          renderCallback={renderCallback}
          destroyCallback={destroyCallback}
        />
        {this.renderConnectors()}
      </div>
    );
  }

  public shouldComponentUpdate = (nextProps: NodeProps<NodeType>): boolean => {
    return nextProps.diagramMakerNode !== this.props.diagramMakerNode;
  }

  private getConnectors(): ConnectorProps[] {
    const { id, diagramMakerData } = this.props.diagramMakerNode;
    const { connectorPlacement } = this.props;
    const { width, height } = diagramMakerData.size;
    const horizontalCenter = width / 2;
    const verticalCenter = height / 2;
    const { INPUT, OUTPUT } = ConnectorType;

    switch (connectorPlacement) {
      case ConnectorPlacement.LEFT_RIGHT:
        return [
          { id, position: { x: 0, y: verticalCenter }, type: INPUT },
          { id, position: { x: width, y: verticalCenter }, type: OUTPUT }
        ];
      case ConnectorPlacement.TOP_BOTTOM:
        return [
          { id, position: { x: horizontalCenter, y: 0 }, type: INPUT },
          { id, position: { x: horizontalCenter, y: height }, type: OUTPUT }
        ];
      default:
        return [];
    }
  }

  private getFilteredConnectors(): ConnectorProps[] {
    const connectorProps = this.getConnectors();

    if (!this.props.visibleConnectorTypes) {
      return connectorProps;
    }

    const { INPUT, OUTPUT } = ConnectorType;
    switch (this.props.visibleConnectorTypes) {
      case VisibleConnectorTypes.INPUT_ONLY:
        return connectorProps.filter(connectorProp => connectorProp.type === INPUT);
      case VisibleConnectorTypes.OUTPUT_ONLY:
        return connectorProps.filter(connectorProp => connectorProp.type === OUTPUT);
      case VisibleConnectorTypes.NONE:
        return [];
    }
    return connectorProps;
  }

  private renderConnectors(): JSX.Element[] {
    return this.getFilteredConnectors().map((connector, i) => {
      const { id, type, position } = connector;
      return <Connector id={id} key={i} type={type} position={position} />;
    });
  }
}
