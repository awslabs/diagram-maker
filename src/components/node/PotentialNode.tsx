import * as Preact from 'preact';

import { ComposeView } from 'diagramMaker/components/common';
import { BoundRenderCallback, DestroyCallback } from 'diagramMaker/service/ConfigService';
import { DiagramMakerComponentsType } from 'diagramMaker/service/ui/types';
import { Position, Size } from 'diagramMaker/state/types';

import './Node.scss';

export interface PotentialNodeProps {
  typeId: string;
  position: Position;
  size: Size;
  renderCallback: BoundRenderCallback;
  destroyCallback: DestroyCallback;
}

export default class PotentialNode extends Preact.Component<PotentialNodeProps, {}> {

  public render(): JSX.Element {
    const { typeId, position, renderCallback, destroyCallback, size } = this.props;
    const { x, y } = position;
    const { width, height } = size;
    const transform = `translate3d(${x}px, ${y}px, 0)`;

    return (
      <div
        className="dm-potential-node"
        style={{ width, height, transform }}
        data-id={typeId}
        data-type={DiagramMakerComponentsType.POTENTIAL_NODE}
        data-event-target={true}
      >
        <ComposeView
          renderCallback={renderCallback}
          destroyCallback={destroyCallback}
        />
      </div>
    );
  }
}
