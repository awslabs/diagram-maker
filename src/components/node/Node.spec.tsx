import { Draft } from 'immer';
import * as Preact from 'preact';
import { shallow } from 'preact-render-spy';

import { Node } from 'diagramMaker/components/node';
import { ConnectorPlacement, VisibleConnectorTypes } from 'diagramMaker/service/ConfigService';
import { DiagramMakerNode } from 'diagramMaker/state/types';

describe('Node', () => {
  const destroyCallback = jest.fn();
  const renderCallback = jest.fn((container: HTMLElement): HTMLElement => {
    const element = document.createElement('div');
    element.textContent = 'Example node content';
    container.appendChild(element);
    return element;
  });

  const getDiagramMakerNode = (): Draft<DiagramMakerNode<void>> => ({
    diagramMakerData: {
      position: {
        x: 400,
        y: 600
      },
      size: {
        height: 300,
        width: 200
      }
    },
    id: 'myNode'
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders at the given position and size', () => {
    const node = shallow(
      <Node
        diagramMakerNode={getDiagramMakerNode()}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(node).toMatchSnapshot();
  });

  it('renders without any connectors', () => {
    const node = shallow(
      <Node
        diagramMakerNode={getDiagramMakerNode()}
        renderCallback={renderCallback}
        destroyCallback={destroyCallback}
      />
    );

    expect(node).toMatchSnapshot();
  });

  it('renders connectors on the top and bottom', () => {
    const node = shallow(
      <Node
        diagramMakerNode={getDiagramMakerNode()}
        renderCallback={renderCallback}
        connectorPlacement={ConnectorPlacement.TOP_BOTTOM}
        destroyCallback={destroyCallback}
      />
    );

    expect(node).toMatchSnapshot();
  });

  it('renders connectors on the left and right', () => {
    const node = shallow(
      <Node
        diagramMakerNode={getDiagramMakerNode()}
        renderCallback={renderCallback}
        connectorPlacement={ConnectorPlacement.LEFT_RIGHT}
        destroyCallback={destroyCallback}
      />
    );

    expect(node).toMatchSnapshot();
  });

  describe('visibleConnectorTypes', () => {
    it('renders connector only for output', () => {
      const diagramMakerNode = getDiagramMakerNode();
      const visibleConnectorTypes = VisibleConnectorTypes.OUTPUT_ONLY;
      const node = shallow(
        <Node
          diagramMakerNode={diagramMakerNode}
          renderCallback={renderCallback}
          connectorPlacement={ConnectorPlacement.LEFT_RIGHT}
          destroyCallback={destroyCallback}
          visibleConnectorTypes={visibleConnectorTypes}
        />
      );

      expect(node).toMatchSnapshot();
    });

    it('renders connector only for input', () => {
      const diagramMakerNode = getDiagramMakerNode();
      const visibleConnectorTypes = VisibleConnectorTypes.INPUT_ONLY;
      const node = shallow(
        <Node
          diagramMakerNode={diagramMakerNode}
          renderCallback={renderCallback}
          connectorPlacement={ConnectorPlacement.LEFT_RIGHT}
          destroyCallback={destroyCallback}
          visibleConnectorTypes={visibleConnectorTypes}
        />
      );

      expect(node).toMatchSnapshot();
    });

    it('renders no connectors', () => {
      const diagramMakerNode = getDiagramMakerNode();
      const visibleConnectorTypes = VisibleConnectorTypes.NONE;
      const node = shallow(
        <Node
          diagramMakerNode={diagramMakerNode}
          renderCallback={renderCallback}
          connectorPlacement={ConnectorPlacement.LEFT_RIGHT}
          destroyCallback={destroyCallback}
          visibleConnectorTypes={visibleConnectorTypes}
        />
      );

      expect(node).toMatchSnapshot();
    });

    it('renders input & output connectors', () => {
      const diagramMakerNode = getDiagramMakerNode();
      const visibleConnectorTypes = VisibleConnectorTypes.BOTH;
      const node = shallow(
        <Node
          diagramMakerNode={diagramMakerNode}
          renderCallback={renderCallback}
          connectorPlacement={ConnectorPlacement.LEFT_RIGHT}
          destroyCallback={destroyCallback}
          visibleConnectorTypes={visibleConnectorTypes}
        />
      );

      expect(node).toMatchSnapshot();
    });
  });

  describe('shouldComponentUpdate', () => {
    it('returns false if nextProps and the current props point to the same object', () => {
      const node = shallow(
        <Node
          diagramMakerNode={getDiagramMakerNode()}
          renderCallback={renderCallback}
          destroyCallback={destroyCallback}
        />
      );

      const component = node.component();
      const { shouldComponentUpdate, props } = component;

      expect(shouldComponentUpdate(props)).toBe(false);
    });

    it('returns true if nextProps and the current props point to a different object', () => {
      const node = shallow(
        <Node
          diagramMakerNode={getDiagramMakerNode()}
          renderCallback={renderCallback}
          destroyCallback={destroyCallback}
        />
      );

      const component = node.component();
      const { shouldComponentUpdate, props } = component;
      const newProps = { ...props, diagramMakerNode: getDiagramMakerNode() };

      expect(shouldComponentUpdate(newProps)).toBe(true);
    });
  });
});
