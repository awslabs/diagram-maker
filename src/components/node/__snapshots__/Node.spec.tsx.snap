// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node renders at the given position and size 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="dm-node"
  style="width: 200px; height: 300px; transform: translate3d(400px, 600px, 0);"
  data-id="myNode"
  data-type="DiagramMaker.Node"
  data-event-target={true}
  data-draggable={true}
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Node renders connectors on the left and right 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="dm-node"
  style="width: 200px; height: 300px; transform: translate3d(400px, 600px, 0);"
  data-id="myNode"
  data-type="DiagramMaker.Node"
  data-event-target={true}
  data-draggable={true}
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
  <Connector
    id="myNode"
    type={0}
    position={
      Object {
        "x": 0,
        "y": 150
      }
    }
  >
  </Connector>
  <Connector
    id="myNode"
    type={1}
    position={
      Object {
        "x": 200,
        "y": 150
      }
    }
  >
  </Connector>
</div>

`;

exports[`Node renders connectors on the top and bottom 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="dm-node"
  style="width: 200px; height: 300px; transform: translate3d(400px, 600px, 0);"
  data-id="myNode"
  data-type="DiagramMaker.Node"
  data-event-target={true}
  data-draggable={true}
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
  <Connector
    id="myNode"
    type={0}
    position={
      Object {
        "x": 100,
        "y": 0
      }
    }
  >
  </Connector>
  <Connector
    id="myNode"
    type={1}
    position={
      Object {
        "x": 100,
        "y": 300
      }
    }
  >
  </Connector>
</div>

`;

exports[`Node renders without any connectors 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="dm-node"
  style="width: 200px; height: 300px; transform: translate3d(400px, 600px, 0);"
  data-id="myNode"
  data-type="DiagramMaker.Node"
  data-event-target={true}
  data-draggable={true}
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;

exports[`Node visibleConnectorTypes renders connector only for input 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="dm-node"
  style="width: 200px; height: 300px; transform: translate3d(400px, 600px, 0);"
  data-id="myNode"
  data-type="DiagramMaker.Node"
  data-event-target={true}
  data-draggable={true}
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
  <Connector
    id="myNode"
    type={0}
    position={
      Object {
        "x": 0,
        "y": 150
      }
    }
  >
  </Connector>
</div>

`;

exports[`Node visibleConnectorTypes renders connector only for output 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="dm-node"
  style="width: 200px; height: 300px; transform: translate3d(400px, 600px, 0);"
  data-id="myNode"
  data-type="DiagramMaker.Node"
  data-event-target={true}
  data-draggable={true}
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
  <Connector
    id="myNode"
    type={1}
    position={
      Object {
        "x": 200,
        "y": 150
      }
    }
  >
  </Connector>
</div>

`;

exports[`Node visibleConnectorTypes renders input & output connectors 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="dm-node"
  style="width: 200px; height: 300px; transform: translate3d(400px, 600px, 0);"
  data-id="myNode"
  data-type="DiagramMaker.Node"
  data-event-target={true}
  data-draggable={true}
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
  <Connector
    id="myNode"
    type={0}
    position={
      Object {
        "x": 0,
        "y": 150
      }
    }
  >
  </Connector>
  <Connector
    id="myNode"
    type={1}
    position={
      Object {
        "x": 200,
        "y": 150
      }
    }
  >
  </Connector>
</div>

`;

exports[`Node visibleConnectorTypes renders no connectors 1`] = `
preact-render-spy (1 nodes)
-------
<div
  class="dm-node"
  style="width: 200px; height: 300px; transform: translate3d(400px, 600px, 0);"
  data-id="myNode"
  data-type="DiagramMaker.Node"
  data-event-target={true}
  data-draggable={true}
>
  <ComposeView
    renderCallback={[Function mockConstructor]}
    destroyCallback={[Function mockConstructor]}
  >
  </ComposeView>
</div>

`;
